sky/backends/cloud_vm_backend.py

```diff
@@ -0,0 +1,351 @@
+        tags={},
+        resume_stopped_nodes=True)
+
+    fh = logging.FileHandler(log_abs_path)
+    fh.setLevel(logging.DEBUG)
+    try:
+        logger.addHandler(fh)
+        logger.debug(_TITLE.format('Provisioning'))
+        logger.debug('Provision config:\n'
+                     f'{json.dumps(bootstrap_config.dict(), indent=2)}')
+        return _bulk_provision(cloud, region, zones, cluster_name,
+                               bootstrap_config)
+    except Exception:  # pylint: disable=broad-except
+        logger.exception(f'Provision cluster {cluster_name} failed.')
+        logger.error('*** Failed provisioning the cluster. ***')
+
+        # If cluster was previously UP or STOPPED, stop it; otherwise
+        # terminate.
+        # FIXME(zongheng): terminating a potentially live cluster is
+        # scary. Say: users have an existing cluster that got into INIT, do
+        # sky launch, somehow failed, then we may be terminating it here.
+        terminate = not is_prev_cluster_healthy
+        terminate_str = ('Terminating' if terminate else 'Stopping')
+        logger.error(f'*** {terminate_str} the failed cluster. ***')
+        # TODO(suquark): In the future we should not wait for cluster stopping
+        #  or termination. This could speed up fail over quite a lot.
+        teardown_cluster(repr(cloud),
+                         region.name,
+                         cluster_name,
+                         terminate=terminate)
+        return None
+    finally:
+        logger.removeHandler(fh)
+        fh.close()
+
+
+def teardown_cluster(cloud_name: str, region: str, cluster_name: str,
+                     terminate: bool) -> None:
+    """Deleting or stopping a cluster."""
+    if terminate:
+        provision.terminate_instances(cloud_name, region, cluster_name)
+        provision_utils.remove_cluster_profile(cluster_name)
+    else:
+        provision.stop_instances(cloud_name, region, cluster_name)
+    try:
+        teardown_verb = 'Terminating' if terminate else 'Stopping'
+        with backend_utils.safe_console_status(
+                f'[bold cyan]{teardown_verb} '
+                f'[green]{cluster_name}\n'
+                f'[white] Press Ctrl+C to send the task to background.'):
+            if terminate:
+                provision.wait_instances(cloud_name, region, cluster_name,
+                                         'terminated')
+            else:
+                provision.wait_instances(cloud_name, region, cluster_name,
+                                         'stopped')
+
+    except KeyboardInterrupt:
+        pass
+
+
+def _post_provision_setup(
+        cloud_name: str, cluster_name: str, cluster_yaml: str,
+        local_wheel_path: pathlib.Path, wheel_hash: str,
+        provision_metadata: provision_comm.ProvisionMetadata,
+        custom_resource: Optional[str]) -> provision_comm.ClusterMetadata:
+    # TODO(suquark): in the future, we only need to mount credentials
+    #  for controllers.
```

Comment:
This may not be true as the cluster need to access the cloud bucket with the credentials.